File: c/92.c
#include <stdio.h>

int sos(int n)
{
	int sum = 0;
	do sum += (n % 10) * (n % 10); while (n /= 10);
	return sum;
}

int arrive89(int n)
{
	while (n != 1 && n != 89)
		n = sos(n);
	return n == 89;
}

int main(void)
{
	int i = 0, ns = 0;
	while (++i < 1e5)
		ns += arrive89(i);

	printf("%d\n", ns);
	return 0;
}

File: hs/92.hs
import Data.Char (ord)

sos = sum . map (\c -> (ord c - 48) * (ord c - 48)) . show

arrive89  1 = False
arrive89 89 = True
arrive89  n = arrive89 (sos n)

main = print $ length $ filter arrive89 [1..10000000]
